backend_new/src/services/application-csv-export.service.ts

```diff
@@ -0,0 +1,596 @@
+import fs from 'fs';
+import path from 'path';
+import { Inject, Injectable, Scope } from '@nestjs/common';
+import { REQUEST } from '@nestjs/core';
+import { Request as ExpressRequest } from 'express';
+import { view } from './application.service';
+import { PrismaService } from './prisma.service';
+import { MultiselectQuestionService } from './multiselect-question.service';
+import { ApplicationCsvQueryParams } from '../dtos/applications/application-csv-query-params.dto';
+import { SuccessDTO } from '../dtos/shared/success.dto';
+import { UnitType } from '../dtos/unit-types/unit-type.dto';
+import { Address } from '../dtos/addresses/address.dto';
+import { ApplicationMultiselectQuestion } from 'src/dtos/applications/application-multiselect-question.dto';
+import MultiselectQuestion from '../dtos/multiselect-questions/multiselect-question.dto';
+import { ApplicationFlaggedSet } from '../dtos/application-flagged-sets/application-flagged-set.dto';
+import { UnitTypeService } from './unit-type.service';
+
+view.csv = {
+  ...view.details,
+  applicationFlaggedSet: true,
+  listings: false,
+};
+
+type CsvHeader = {
+  path: string;
+  label: string;
+  format?: (val: unknown) => unknown;
+};
+
+@Injectable({ scope: Scope.REQUEST })
+export class ApplicationCsvExporterService {
+  constructor(
+    @Inject(REQUEST) private req: ExpressRequest,
+    private prisma: PrismaService,
+    private multiselectQuestionService: MultiselectQuestionService,
+    private unitTypeService: UnitTypeService,
+  ) {}
+
+  /*
+   * Prepares an export and streams file
+   */
+  async export(queryParams: ApplicationCsvQueryParams): Promise<SuccessDTO> {
+    await this.authorizeCSVExport(this.req.user, queryParams.listingId);
+
+    if (queryParams.includeDemographics) {
+      view.csv.demographics = true;
+    }
+
+    const applications = await this.prisma.applications.findMany({
+      include: view.csv,
+      where: {
+        listingId: queryParams.listingId,
+      },
+    });
+
+    // get all multiselect questions for a listing to build csv headers
+    const multiSelectQuestions =
+      await this.multiselectQuestionService.findByListingId(
```

Comment:
this is something I'm will to cede on, but I think my preference is:

if we aren't using very complex logic already built into a service, we can just query for the items in the service we're looking at.

What I'm trying to get at it is: if this is the only thing we are using the `multiselectQuestionService` for, we can make a function called `findMultiselectQuestionsByListingId()` in this file and not need to import a whole service for a relatively small amount of logic.


like I said though, I'm open to a counter argument though!

Reply:
I was ok with moving `unitTypeToReadable` into this file, because it's more of a formatting function, but structurally, I think it's overall better to keep functions like this one within the service.

If at some point another feature requires the use of finding multiselect questions by listing Id, if this function is within application-csv-export then another dev may not know about it and we'll end up with duplicate code. I think straying away from doing that so that we have less dependencies makes the code more like spaghetti, because you can't be sure where a function might be defined.

Reply:
Ah yeah that makes a ton of sense! thanks!